File: 328-odd-even-linked-list/odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None
        
        p1 = head
        p2 = head.next
        odd = p1
        even = p2

        while p1 and p2 and p2.next:
            p1.next  = p2.next
            p2.next = p2.next.next
            p1 = p1.next
            p2 = p2.next
        
        p1.next = even
        return odd

File: 328-odd-even-linked-list/odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if(head == null){
            return null;
        }
        ListNode odd = new ListNode(1); // dummy odd node
        ListNode even = new ListNode(0); // dummy even node

        ListNode p = head;
        ListNode t_o = odd;
        ListNode t_e = even;
        int count = 1;
        while (p!=null){
            boolean isOdd = ((count)%2==1);
            ListNode n = new ListNode(p.val);
            if (isOdd){
                t_o.next = n;
                t_o = t_o.next;

            }else{
                t_e.next = n;
                t_e = t_e.next;
            }
            p = p.next;
            count++;
        }
        t_o.next = even.next;
        return odd.next;
        
    }
}